File: Implementations/C/lcs.c
#include <stdio.h>
#include <string.h>
#define MX 501
int b[MX][MX], c[MX][MX], i, j, m, n;

void lcs_len(char* x, char* y) {
    m = strlen(x)-1, n = strlen(y)-1;
    for(i = 0; i<=m; i++) c[i][0] = 0;
    for(i = 1; i<=n; i++) c[0][i] = 0;

    for(i = 1; i<=m; i++)
        for(j = 1; j<=n; j++) {
            if(x[i] == y[j]) {
                c[i][j] = c[i-1][j-1]+1;
                b[i][j] = 1;
            } else if(c[i-1][j] >= c[i][j-1]) {
                c[i][j] = c[i-1][j];
                b[i][j] = 2;
            } else {
                c[i][j] = c[i][j-1];
                b[i][j] = 3;
            }
        }
}

void printOut(int m, int n, char* x) {
    if(m==0||n==0)return;
    if(b[m][n]==1) {
        printOut(m-1,n-1,x);
        printf("%c", x[m]);
    } else if(b[m][n] ==2) printOut(m-1,n,x);
    else printOut(m,n-1,x);
}

int main() {
    printf("Enter two strings : ");
    char x[MX], y[MX];
    x[0] = y[0] = '0';
    scanf("%s %s", x+1,y+1);
    lcs_len(x, y);
    printf("%d\n", c[m][n]);
    printOut(m, n, x);
    return 0;
}

File: Implementations/Python/lcs.py
import timeit


def lcs(u, v):
    m, n = len(u), len(v)
    b, c = [[0 for i in range(n + 1)] for j in range(m + 1)], [[0 for i in range(n + 1)] for j in range(m + 1)]
    for i in range(1, m):
        for j in range(1, n):
            if u[i] == v[j]:
                c[i][j], b[i][j] = c[i - 1][j - 1] + 1, 1
            elif c[i - 1][j] >= c[i][j - 1]:
                c[i][j], b[i][j] = c[i - 1][j], 2
            else:
                c[i][j], b[i][j] = c[i][j - 1], 3
    return c, b


def printOut(b, s, i, j):
    if i == 0 or j == 0:
        return
    if b[i][j] == 1:
        printOut(b, s, i - 1, j - 1)
        print(s[i], end='')
    elif b[i][j] == 2:
        printOut(b, s, i - 1, j)
    else:
        printOut(b, s, i, j - 1)


if __name__ == '__main__':
    print("Enter two strings in one line :")
    d, e = [str(i) for i in input().split()]
    d, e = '0' + d, '0' + e
    t = timeit.default_timer()
    p, q = lcs(d, e)
    t = timeit.default_timer() - t
    print("Answer : ", p[len(d) - 1][len(e) - 1])
    printOut(q, d, len(d) - 1, len(e) - 1)
    print("\n Time elapsed : ", t)
